content/posts/messaging-infrastructure.md

```diff
@@ -0,0 +1,96 @@
+
+## Consistent messaging
+
+As we mentioned in [our first post](https://exactly-once.github.io/posts/consistent-messaging/), the message processing needs to follow a specific pattern in order for the distributed system to be consistent. We called that pattern *consistent messaging*. Let us now examine possible ways of arranging the basic operations of message processing into transactions such that fit the consistent messaging pattern.
+
+## No transactions
+
+The trivial approach is to not use any transactions at all. Each operation is executed in its own context. 
+
+{{< figure src="/posts/messaging-infrastrcture-no-transactions.png" title="No transactions">}}
+
+In this case, we have three distinct failure modes:
+- business logic has executed
+- state change has been persisted
+- outgoing messages have been sent
+
+If we are sending out more than one message, the number of possible failure scenarios increases because sending each message constitutes a separate operation.
+
+This approach requires the least support from the messaging infrastructure but is the hardest one to work with as the business logic needs to take into account all these failure modes and support deterministic behavior during retires. As we proved in [consistent messaging](https://exactly-once.github.io/posts/consistent-messaging/) this is all but a trivial task even for the simplest business logic.
+
+## All-or-nothing
+
+At the opposite end of the spectrum we have the all-or-nothing approach, sometimes referred to as *fully-consistent*. In this mode a single atomic and durable transaction spans all the operations excluding execution of the business logic.
+
+{{< figure src="/posts/messaging-infrastrcture-all-or-nothing.png" title="All or nothing">}}
+
+The all-or-nothing mode is guaranteed to not generate duplicate messages in the course of the communication as the incoming message is consumed only if the outgoing messages are sent out, and vice versa. So, given there is no duplicate messages, does it mean this is equivalent to *exactly-once message delivery*? Of course the answer is no. 
+
+The fact that exactly-once message delivery is impossible can be demonstrated using the famous two generals problem [^1]. In short, the two parties (in our case the messaging infrastructure and the CPU) need to agree what is the state of message processing. This is proven impossible if these two parties don't share state. You might now wonder what would happen if they did? Imagine a computer which main memory and registers (including the program counter -- PC) are backed up by a durable storage. Such a computer could implement exactly-once message delivery mode by storing a snapshot of memory and registers for each processed message.
```

Comment:
> This is proven impossible if these two parties don't share state. You might now wonder what would happen if they did? Imagine a computer which main memory and registers (including the program counter -- PC) are backed up by a durable storage. Such a computer could implement exactly-once message delivery mode by storing a snapshot of memory and registers for each processed message.

I would be a bit more specific about the "state sharing" part and say that this could work only if the process, broker, the storage and all the communication links would be a single failure domain. Otherwise, we are back in the distributed systems world and 2 generals hold.

Reply:
Maybe let's skip this whole para? It is a bit off-topic anyway?